ls1intum/Artemis
# [Feature] Exercise Performance Visualisation

### Checklist

### Motivation and Context


This pull request introduces a new visualization for students, which they can use to compare their own performance in a course with the performance of the rest of the course. 

### Description

Using the recently introduced Participant Score table, we efficiently calculate the logged-in student's performance, the average performance of the course, and the best performance of a student for all assignments of a course. This information is then sent to the client and displayed in a Chart.js graph (see screenshot section)

Note: We only consider released and finished (assessment due date over) exercises!
### Steps for Testing


1. Log in to Artemis with a student who has some scores in a course
2. Open the course overview page for a course with some scores
3. Scroll down and you should see the exercise scores chart

### Test Coverage



exercise-scores-chart.component.ts 73% **(improving further will be very hard in a unit test as I would need to simulate a user clicking specific points on the chart, effort would be not reasonable imo)**
exercise-scores-chart.service.ts 88%
ExerciseScoresChartService.java 97%
ExerciseScoresChartResource.java 100%



### Screenshots



![image](https://user-images.githubusercontent.com/29718932/114269747-e659e900-9a08-11eb-928a-952f7558d383.png)

https://user-images.githubusercontent.com/29718932/114269697-bca0c200-9a08-11eb-861c-ceb80eb792ad.mov
